lib/ZumoOta/src/BootloaderCom.cpp

```diff
@@ -97,26 +102,122 @@ class IntProgCmdProvider: public CmdProvider {
 private:
     static const uint32_t SEQ_LENGHT = 12; /**< Length of the command sequence */
     uint32_t m_index; /**< current index in sequence */
-    static const CommandInfo m_cmds[SEQ_LENGHT]; /**< Array of command information */
-    static const ResponseInfo m_responses[SEQ_LENGHT]; /**< Array of response information */
+    static  CommandInfo m_cmds[SEQ_LENGHT]; /**< Array of command information */
+    static  ResponseInfo m_responses[SEQ_LENGHT]; /**< Array of response information */
 };
 
-#if 0
+/**
+ * @class FwProgCmdProvider
+ * @brief Implementation of the command provider for firmware programming.
+ *
+ * This class provides commands for firmware programming, specifically tailored
+ * for the Zumo32U4 device.
+ */
 class FwProgCmdProvider : public CmdProvider {
     public:
-        FwProgCmdProvider(FileHandle * fs);
-}
-#endif
+        /**
+         * @brief Constructor for the FwProgCmdProvider class.
+         * @param fileName The name of the file to be used for firmware programming.
+         * Initializes a new instance of the FwProgCmdProvider class with the provided file name.
+         * @remarks The file specified by fileName should contain the firmware to be programmed.
+         */
+         FwProgCmdProvider(const std::string& fileName):
+            m_index(0),
+            m_fileName(fileName)
+    {}
+
+    /**
+     * @brief Resets the command provider to its initial state.
+     *
+     * This method resets the internal index that tracks the current command
+     * in the command sequence memory.
+     */
+    virtual void reset() { m_index = 0;}
+
+    /**
+     * @brief Retrieves the next command and the next response.
+     *
+     * This method returns the next command and its associated response
+     * from the command sequence memory.
+     *
+     * @param cmd Reference to a pointer to the next command.
+     * @param rsp Reference to a pointer to the next response.
+     * @return True if there is a next command and next response, false otherwise.
+     */
+    virtual bool next( CommandInfo *& cmd, ResponseInfo *& rsp)
+    {
+        /*Determine the command and response for the current chunk.*/
+        uint32_t commandIndex = m_index % SEQ_LENGHT;
+        cmd = &m_cmds[commandIndex];
+
+        if(cmd->command == Zumo32U4Specification::SET_MEMORY_ADDR)
+        {
+            cmd->commandsize = FinalLength;
+        }
+        else if(cmd->command == Zumo32U4Specification::WRITE_MEMORY_PAGE)
+        {
+            size_t bytesRead = m_fileManager.read128Bytes(m_fileName.c_str(), m_buffer);
+            /*If no bytes were read, there are no more chunks left.*/
+            if (0 == bytesRead)

```

Comment:
What do you do if your firmware ist 129 bytes long ?

You have to handle the fact that the last block is in between 0 and 127 tail bytes.
In this case you need to fill the gap to the missing blocksize with a pattern, usually 0xFF or 0x00,
The value is what the flash reads after an erase, to avoid unnecessary programming of the bits.